task2/rest-api-basics/repository/src/main/java/com/epam/esm/repository/impl/TagRepositoryImpl.java

```diff
@@ -0,0 +1,99 @@
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.core.PreparedStatementCreator;
+import org.springframework.jdbc.support.GeneratedKeyHolder;
+import org.springframework.jdbc.support.KeyHolder;
+import org.springframework.stereotype.Repository;
+
+import java.sql.PreparedStatement;
+import java.util.List;
+
+@Repository
+public class TagRepositoryImpl implements TagRepository {
+    private static final Logger logger = LogManager.getLogger(TagRepositoryImpl.class);
+
+    @Autowired
+    private JdbcTemplate jdbcTemplate;
+
+    private static final String FIND_ALL_QUERY = "SELECT id t_id, name t_name FROM tag";
+    private static final String FIND_BY_ID_QUERY = "SELECT id t_id, name t_name FROM tag WHERE id = ?";
+    private static final String FIND_BY_NAME_QUERY = "SELECT id t_id, name t_name FROM tag WHERE name = ?";
+    private static final String SAVE_QUERY = "INSERT INTO tag (name) VALUES (?)";
+    private static final String DELETE_BY_NAME_QUERY = "DELETE FROM tag WHERE name = ?";
+
+    @Override
+    public List<Tag> findAll() throws RepositoryException {
+        try {
+            return jdbcTemplate.query(FIND_ALL_QUERY, new TagMapper());
+        } catch (DataAccessException e) {
+            throw new RepositoryException(e);
+        }
+    }
+
+    @Override
+    public Tag findById(Long id) throws RepositoryException {
+        try {
+            return jdbcTemplate.queryForObject(FIND_BY_ID_QUERY, new TagMapper(), id);
+        } catch (EmptyResultDataAccessException e) {
+            return null;
+        } catch (DataAccessException e) {
+            throw new RepositoryException(e);
+        }
+    }
+
+    @Override
+    public Tag findByName(String name) throws RepositoryException {
+        try {
+            return jdbcTemplate.queryForObject(FIND_BY_NAME_QUERY, new TagMapper(), name);
+        } catch (EmptyResultDataAccessException e) {
+            return null;
+        } catch (DataAccessException e) {
+            throw new RepositoryException(e);
+        }
+    }
+
+    @Override
+    public Tag save(Tag tag) throws RepositoryException {
+        try {
+            KeyHolder keyHolder = new GeneratedKeyHolder();
+            PreparedStatementCreator preparedStatementCreator = con -> {
+                PreparedStatement preparedStatement = con.prepareStatement(SAVE_QUERY, PreparedStatement.RETURN_GENERATED_KEYS);
+                preparedStatement.setString(1, tag.getName());
+                return preparedStatement;
+            };
+
+            jdbcTemplate.update(preparedStatementCreator, keyHolder);
+
+            // TODO: 11-Jan-21 keys as null exception
+            tag.setId((Long) keyHolder.getKeys().get(TagMapper.SECONDARY_ID));
+
+            return tag;
+        } catch (DataAccessException e) {
+            throw new RepositoryException(e);
+        }
+    }
+
+    @Override
+    public Long deleteByName(String name) throws RepositoryException {
```

Comment:
@igor-zhuravlev Why not tag id. It's unique.

Reply:
@igor-zhuravlev if you pass a name from above, just check against the existing name. If exists, return, but not, throw exception